### ISSUE TITLE
isDef?

### ISSUE THREAD

**Issue by mreinstein** [snabbdom/snabbdom#932]
At the risk of starting a flamewar (I definitely do _not_ want to start one,) I can't help but look at some parts of the code written in typescript and wonder why we are doing this. For example:

```typescript
type NonUndefined<T> = T extends undefined ? never : T

function isUndef (s: any): boolean {
  return s === undefined
}
function isDef<A> (s: A): s is NonUndefined<A> {
  return s !== undefined
}
```

It seems to me like a lot over overhead, compared to doing `if (s === undefined)`. Typescript could be limiting contributor-ship.

Again, I'm not trying to start a fight or demand that we change this. I'm just taking the temperature to see what the general sentiment on this is. I don't know if it's just me that finds it to be over-engineered or if others feel that way too?

**Comment by wclr** [comment]
How do you want to rewrite this?

```ts
    if (isUndef(vnode.text)) {
      if (isDef(oldCh) && isDef(ch)) {
        if (oldCh !== ch) updateChildren(elm, oldCh, ch, insertedVnodeQueue);
      } else if (isDef(ch)) {
        if (isDef(oldVnode.text)) api.setTextContent(elm, "");
        addVnodes(elm, null, ch, 0, ch.length - 1, insertedVnodeQueue);
      } else if (isDef(oldCh)) {
        removeVnodes(elm, oldCh, 0, oldCh.length - 1);
      } else if (isDef(oldVnode.text)) {
        api.setTextContent(elm, "");
      }
    } else if (oldVnode.text !== vnode.text) {
      if (isDef(oldCh)) {
        removeVnodes(elm, oldCh, 0, oldCh.length - 1);
      }
      api.setTextContent(elm, vnode.text!);
    }
```

Functions are used to make code terser and easier to read. As for performance, I don't think that it does really matter.

> Typescript could be limiting contributor-ship.

This is funny. You should learn typescript, this is just lightweight types over js.

**Comment by mightyiam** [comment]
There may be pieces of TypeScript in this codebase that are not as modern as they can be, because they were written for older versions of both TypeScript and JavaScript. Improvements and even pointers are welcome.

**Comment by mightyiam** [comment]
But we're not ditching TypeScript, obviously. It helps with type safety for this project and some of its consumers. They would not be happy about us ditching TypeScript. And, if anything, TypeScript seems quite popular these days. So I can't imagine it deterring contribution more than attracting it.

**Comment by mreinstein** [comment]
> How do you want to rewrite this?

I never suggested that particular code block needs re-writing.


> As for performance, I don't think that it does really matter.

Performance is a good question, something that would need to be actually measured to be a useful point in either the pro/con side.


> this is just lightweight types over js.

lightweight to who? Humans or computer? :)

sample A:

```javascript
if (vnode.text === undefined)
```

sample B:

```typescript
type NonUndefined<T> = T extends undefined ? never : T

function isUndef (s: any): boolean {
  return s === undefined
}
function isDef<A> (s: A): s is NonUndefined<A> {
  return s !== undefined
}

if (isUndef(vnode.text))
```

**Comment by jvanbruegge** [comment]
Your sample A would work the same for typescript.

Typescript is a must have for me, i usually rule out dependencies that dont have proper types

**Comment by mreinstein** [comment]
ok, fair enough. Just raising the question. If people love it then so be it.

**Comment by paldepind** [comment]
Just to offer a bit of history, the use of the `isDef`/`isUndef` predates the migration to TypeScript and doesn't really have anything to do with TypeScript. I've changed the title of the PR to reflect that, as far as I can tell, it is about `isDef` and not TypeScript.

As for why `isDef` was originally introduced (by me): There are many instances in the Snabbdom codebase where we need to access something of a nested structure, where each layer of the structure may not be there. This is one such example:

```js
vnode.data.hook?.update?.(oldVnode, vnode);
```

As you one can see, today we do this by using the optional chaining `?.` syntax. This was changed in #489—before that the pattern that I used looked like this:


```js
if (isDef(i = ch.data) && isDef(i = i.hook) && isDef(i = i.remove)) {
```

For such a line I found the `isDef` function to make things a lot nice to read. For instance it would often change wether or not something could fit on a single line vs taking up three lines. That is why it was introduced into the code base.

However, in the presence of `?.` I think `isDef` is a lot less useful. In fact, I actually mentioned that in a comment on the before mentioned MR (https://github.com/snabbdom/snabbdom/pull/489#issuecomment-558670750) where I suggest removing the `isDef` and `isUndef` function. Even in code like what @wclr I don't think `isDef`/`isUndef` offers a noticeable advantage. It's a matter of taste however.

To summarize, I don't think the `isDef` function is as useful as it once was and I'm also in favour of removing it.

**Comment by mreinstein** [comment]
> I've changed the title of the PR to reflect that, as far as I can tell, it is about isDef and not TypeScript.

Quite the opposite actually. :-)  My point in highlighting `isDef` implementation is not to pick on this particular function call, more to show how some people may find typescript to be a bit unpleasant to read and generally work with. I realize that there is an extreme minority of people that feel this way. I don't expect this to change in the near future. In fact this issue can probably just be closed. It might be good to restore the  title back to it's original intention before closing it though. :-)


> Just to offer a bit of history

That is helpful/interesting, thanks for the context! 👍 


> However, in the presence of ?. I think isDef is a lot less useful

I guess it depends on how this is handled. Assuming we transpile down to a version of javascript that avoids `?.` I think that's a good plan, because optional chaining is still quite new and not widespread enough yet https://caniuse.com/?search=optional%20chain

>  I don't think the isDef function is as useful as it once was and I'm also in favour of removing it.

Assuming we are talking about _not_ shipping optional chaining in the resulting javascript, I agree. This should be opened as a separate isssue so it's not lumped in with my typescript question.

**Comment by paldepind** [comment]
Ok, well, then I'm confused. Because your original post sounds like a complaint about `isDef` and not about TypeScript. For instance you write:

> It seems to me like a lot over overhead, compared to doing `if (s === undefined)`.

But, we could easily do `if (s === undefined)` and still use TypeScript.

**Comment by mreinstein** [comment]
I don't feel like getting into an argument about whether or not my github issue that I opened around questioning the usage of typescript is truly about the questioning the usage of typescript or not.  I'm telling you as it's author, that was the topic. Let's move on. :-)  It's a moot point since everyone here loves and wants it.
